File: Tree-InOrder/inorder-successor/inorder_successor_in_BST.py
def inorderSuccessor(self, root, p):
    if root is None or p is None:
        return None

    if root.val <= p.val:
        return self.inorderSuccessor(root.right, p)
    else:
        left = self.inorderSuccessor(root.left, p)
        if left is None:
            return root
        else:
            return left


def inorderSuccessor(self, root, p):
    if root is None or p is None:
        return None

    successor = None

    while root is not None and root != p:
        if root.val > p.val:
            successor = root
            root = root.left
        else:
            root = root.right

    if root is None:
        return

    if root.right is None:
        return successor

    root = root.right
    while root.left is not None:
        root = root.left

    return root

File: Tree-InOrder/inorder-successor/inorder-successor-in-bst.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

export function inorderSuccessor(
  root: TreeNode | null,
  p: TreeNode | null
): TreeNode | null {
  if (root === null) {
    return null;
  }

  if (root.val < p.val) {
    return inorderSuccessor(root.right, p);
  } else {
    const left = inorderSuccessor(root.left, p);
    return left ? left : root;
  }
}
